File: src/main/java/gargoyle/rpycg/service/KeyConverter.java
package gargoyle.rpycg.service;

import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyCodeCombination;
import javafx.scene.input.KeyCombination;

import java.util.HashSet;
import java.util.Locale;
import java.util.Optional;
import java.util.Set;

public final class KeyConverter {
    private static final KeyCombination.Modifier[] MODIFIERS = new KeyCombination.Modifier[0];
    private static final char PLUS = '+';
    private static final String UNDERSCORE = "_";

    public String toBinding(KeyCodeCombination key) {
        StringBuilder sb = new StringBuilder(32);
        boolean complex = false;
        if (key.getShift() == KeyCombination.ModifierValue.DOWN) {
            sb.append(KeyCode.SHIFT.getName().toLowerCase(Locale.ENGLISH));
            complex = true;
        }
        if (key.getControl() == KeyCombination.ModifierValue.DOWN) {
            if (complex) {
                sb.append(UNDERSCORE);
            }
            sb.append(KeyCode.CONTROL.getName().toLowerCase(Locale.ENGLISH));
            complex = true;
        }
        if (key.getAlt() == KeyCombination.ModifierValue.DOWN) {
            if (complex) {
                sb.append(UNDERSCORE);
            }
            sb.append(KeyCode.ALT.getName().toLowerCase(Locale.ENGLISH));
            complex = true;
        }
        if (key.getMeta() == KeyCombination.ModifierValue.DOWN) {
            if (complex) {
                sb.append(UNDERSCORE);
            }
            sb.append(KeyCode.META.getName().toLowerCase(Locale.ENGLISH));
            complex = true;
        }
        if (complex) {
            sb.append(UNDERSCORE);
        }
        sb.append(key.getCode().getName().toUpperCase(Locale.ENGLISH));
        return sb.toString();
    }

    public Optional<KeyCodeCombination> toCombination(String keys) {
        try {
            KeyCode keyCode = KeyCode.UNDEFINED;
            Set<KeyCombination.Modifier> modifiers = new HashSet<>(4);
            for (String name : keys.split(UNDERSCORE)) {
                KeyCode value = KeyCode.valueOf(name);
                switch (value) {
                    case SHIFT:
                        modifiers.add(KeyCombination.SHIFT_DOWN);
                        break;
                    case CONTROL:
                        modifiers.add(KeyCombination.CONTROL_DOWN);
                        break;
                    case ALT:
                        modifiers.add(KeyCombination.ALT_DOWN);
                        break;
                    case META:
                        modifiers.add(KeyCombination.META_DOWN);
                        break;
                    default:
                        keyCode = value;
                }
            }
            return Optional.of(new KeyCodeCombination(keyCode, modifiers.toArray(MODIFIERS)));
        } catch (IllegalArgumentException e) {
            return Optional.empty();
        }
    }

    public String toString(KeyCodeCombination keyCodeCombination) {
        StringBuilder sb = new StringBuilder(32);
        boolean complex = false;
        if (keyCodeCombination.getShift() == KeyCombination.ModifierValue.DOWN) {
            sb.append(KeyCode.SHIFT.getName().toUpperCase(Locale.ENGLISH));
            complex = true;
        }
        if (keyCodeCombination.getControl() == KeyCombination.ModifierValue.DOWN) {
            if (complex) {
                sb.append(PLUS);
            }
            sb.append(KeyCode.CONTROL.getName().toUpperCase(Locale.ENGLISH));
            complex = true;
        }
        if (keyCodeCombination.getAlt() == KeyCombination.ModifierValue.DOWN) {
            if (complex) {
                sb.append(PLUS);
            }
            sb.append(KeyCode.ALT.getName().toUpperCase(Locale.ENGLISH));
            complex = true;
        }
        if (keyCodeCombination.getMeta() == KeyCombination.ModifierValue.DOWN) {
            if (complex) {
                sb.append(PLUS);
            }
            sb.append(KeyCode.META.getName().toUpperCase(Locale.ENGLISH));
            complex = true;
        }
        if (complex) {
            sb.append(PLUS);
        }
        sb.append(keyCodeCombination.getCode().getName().toUpperCase(Locale.ENGLISH));
        return sb.toString();
    }
}

File: app/src/main/kotlin/gargoyle/rpycg/service/KeyConverter.kt
package gargoyle.rpycg.service

import javafx.scene.input.KeyCode
import javafx.scene.input.KeyCodeCombination
import javafx.scene.input.KeyCombination

class KeyConverter {
    fun toBinding(key: KeyCodeCombination): String =
        with(mutableListOf<String>()) {
            if (key.shift == KeyCombination.ModifierValue.DOWN) this += KeyCode.SHIFT.getName().lowercase()
            if (key.control == KeyCombination.ModifierValue.DOWN) this += KeyCode.CONTROL.getName().lowercase()
            if (key.alt == KeyCombination.ModifierValue.DOWN) this += KeyCode.ALT.getName().lowercase()
            if (key.meta == KeyCombination.ModifierValue.DOWN) this += KeyCode.META.getName().lowercase()
            this += key.code.getName().uppercase()
            this
        }.joinToString(UNDERSCORE)

    fun toCombination(keys: String): KeyCodeCombination? =
        try {
            var keyCode: KeyCode = KeyCode.UNDEFINED
            val modifiers: MutableList<KeyCombination.Modifier> = mutableListOf()
            for (name in keys.split(UNDERSCORE)) {
                when (val value = KeyCode.valueOf(name)) {
                    KeyCode.SHIFT -> modifiers += KeyCombination.SHIFT_DOWN
                    KeyCode.CONTROL -> modifiers += KeyCombination.CONTROL_DOWN
                    KeyCode.ALT -> modifiers += KeyCombination.ALT_DOWN
                    KeyCode.META -> modifiers += KeyCombination.META_DOWN
                    else -> keyCode = value
                }
            }
            KeyCodeCombination(keyCode, *modifiers.toTypedArray())
        } catch (e: IllegalArgumentException) {
            null
        }

    fun toString(key: KeyCodeCombination): String =
        with(mutableListOf<String>()) {
            if (key.shift == KeyCombination.ModifierValue.DOWN) this += (KeyCode.SHIFT.getName().uppercase())
            if (key.control == KeyCombination.ModifierValue.DOWN) this += (KeyCode.CONTROL.getName().uppercase())
            if (key.alt == KeyCombination.ModifierValue.DOWN) this += (KeyCode.ALT.getName().uppercase())
            if (key.meta == KeyCombination.ModifierValue.DOWN) this += (KeyCode.META.getName().uppercase())
            this += (key.code.getName().uppercase())
            this
        }.joinToString(PLUS)

    companion object {
        private const val PLUS = "+"
        private const val UNDERSCORE = "_"
    }
}
